Constestant and Loop - Setup Domain Entity

StoyanGoranov opened the pull request:
Related to: https://github.com/Not-Endurance/not-timing-system/issues/191

StoyanGoranov committed:
add Contestant class to Domain Setup

https://github.com/Not-Endurance/not-timing-system/issues/191

StoyanGoranov committed:
add ContestantFormModel

https://github.com/Not-Endurance/not-timing-system/issues/191

StoyanGoranov committed:
add ContestantGrid component and persistance

Modifies the Competition Domain Entity, NTS.Persistance.State to include
a state of the Competition parent, adds a ContesntantRepository and an
implementation of the CompetitionChildrenBehind

https://github.com/Not-Endurance/not-timing-system/issues/191

StoyanGoranov commented on src-v2/Judge/NTS.Judge/Events/CompetitionChildrenBehind.cs:
```diff
@@ -0,0 +1,45 @@
+    private readonly IParentRepository<Contestant> _contestantRepository;
+    private Competition _competition;
+
+    public CompetitionChildrenBehind(IParentRepository<Contestant> contestants)
```
The following feels like it must be conceptually wrong but I seem to lack the knowledge to reason why.
 ```C#
   public CompetitionChildrenBehind(IRepository<Competition> competition, IParentRepository<Contestant> contestants, Competition competitionParent)
   {
       _competitionRepository = competition;
       _contestantRepository = contestants;
   }
```

StoyanGoranov committed:
Modify MudValidationInjector

Use MudFormComponent as common ancestor and remove Create<T> overload

StoyanGoranov committed:
Configure Setup Boundary Entities

Finish the structure of the Setup Boundary Domain Entities as per the
diagram

https://github.com/Not-Endurance/not-timing-system/issues/170

StoyanGoranov committed:
something is missing

Commit:
Commit INotBehindWithChildren implementation

StoyanGoranov committed:
Add INotBehindWithChildren, Revamp Competition Component, Add ContestantPage in Competition

StoyanGoranov committed:
fix: duplicate CONTESTANT_UPDATE route

StoyanGoranov committed:
Add fixes to Competition DomainEntity

Add a ctor that accepts and Id to ensure that operations over the same
data do not create separate instances of the class;
Modify the parameter names of the JsonConstructor to ensure successful
desearilization from string to Competition instance.

https://github.com/Not-Endurance/not-timing-system/pull/195

StoyanGoranov committed:
Inherit ISingletonService at INotBehindWithChildren

The change ensure the interface that governs children of entities is
registered as Singleton service at Not.Injection.ConventionExtensions

https://github.com/Not-Endurance/not-timing-system/pull/195

StoyanGoranov committed:
Add Id parameter to Contestant.cs methods

The Id is needed to ensure unneccesary instance of the entity aren't
created on Update and Json deserialization

https://github.com/Not-Endurance/not-timing-system/pull/195

StoyanGoranov committed:
Update Competition and Contestant Components

The latest structure allows adding Contestant children to a Competition
when updating a Competition instance.

https://github.com/Not-Endurance/not-timing-system/pull/195

achobanov committed:
Fix CompetitionChildrenBehind and repositories

StoyanGoranov committed:
Fix TimePicker logic at ContestantForm.razor

fixes the faulty logic that adds the time picked by the user to a
DataTimeOffset

https://github.com/Not-Endurance/not-timing-system/pull/195

StoyanGoranov committed:
fix formatting - remove empty lines

StoyanGoranov committed:
Fix DomainEntities logic to show validation messages

This change ensures the DomainEntities correctly reference the name of
the concerned property when throwing a DomainException

https://github.com/Not-Endurance/not-timing-system/pull/195

StoyanGoranov committed:
Domain - add CUD logic to Loop and its parent - Competition

Adhere to the hierarchical structure of the DomainEntities by adding
Loop entities as collection of children to Competition

https://github.com/Not-Endurance/not-timing-system/pull/195

StoyanGoranov committed:
Add Loop Behind operations and persistence

The LoopBehind logic belongs to CompetitionChildrenBehind and the
persistence logic to CompetitionRepository because Loop is a child
collection of Competition

https://github.com/Not-Endurance/not-timing-system/pull/195

StoyanGoranov committed:
Add Loop Setup Components

Add LoopForm, LoopFormModel, LoopUpdate and LoopGrid; Utilize the Loop
Grid in the CompetitionUpdate Component

StoyanGoranov committed:
Add validation logic to Loop Setup Domain Entity

Ensure time and distance properties aren't set to zero or less.

https://github.com/Not-Endurance/not-timing-system/pull/195

StoyanGoranov committed:
fix missing Min property at LoopForm MudNumericField

achobanov commented on src-v2/NTS.Persistance/State.cs:
```diff
@@ -7,14 +8,26 @@ public class State :
     public List<Official> Officials => Event?.Officials.ToList() ?? new();
     public List<Competition> Competitions => Event?.Competitions.ToList() ?? new();
 
+    public List<Contestant> Contestants => Competition?.Contestants.ToList() ?? new();
```
Is this correct?

achobanov commented on src-v2/NTS.Persistance/State.cs:
```diff
@@ -7,14 +8,26 @@ public class State :
+
+    public List<Tandem> Tandems { get; } = new();
+
+    public Competition? Competition
+    {
+        get => _competitions.FirstOrDefault();
+        set => _competitions = value == null ? new() : new() { value };
+    }
```
Is this necessary?

achobanov commented on src-v2/Not/Not.Blazor/TM/Navigation/NotNavigator.cs:
```diff
@@ -33,7 +33,7 @@ public T ConsumeParameter<T>()
         }
         var result = _parameters.Get<T>();
-        _parameters = null;
+        //_parameters = null;
```
Should this exist?

achobanov reviewed (changes requested):
There are warnings that should be addressed in Contestant.cs

There are bugs with Loop update and loop create operations. I suggest that you always check the full functionality before submitting a PR for review

achobanov commented:
Furthermore, there was no task and issue created for Loop implementation. I've created #206. You should be committing to the sub-task issues, rather than the big umbrella (#170) as it is too big to provide a useful context when looking at a given commit. Also the PR description should include `closes #` for both subtasks as we want Github to close them automatically as they hit master.

StoyanGoranov commented on src-v2/Judge/NTS.Judge.Blazor/Pages/Setup/Competitions/CompetitionForm.razor:
```diff
@@ -62,6 +62,6 @@
-        return Competition.Update(_model.Name, _model.Type, startTimeOffset);
+        return Competition.Update(_model.Id, _model.Name, _model.Type, startTimeOffset);
     }
 }
```
It isn't since StartDate isn't a property of the Domain Entity but it still seems like it needs to be validated somehow.

StoyanGoranov commented on src-v2/Judge/NTS.Judge.Blazor/Pages/Setup/Contestants/ContestantGrid.razor:
```diff
@@ -0,0 +1,42 @@
+    <NotH2 Class="pt-6 pl-6">@Localizer.Get("Contestants")</NotH2>
+    <NotButtonCreate Class="mt-1 ml-6 mb-3" OnClick="_contestantNavigator.Create" />
+    <MudDivider />
+    <MudDataGrid  T="Contestant" Items="_parentBehind.Children" ReadOnly="true" Bordered="true" Dense="true">
```
ToString is indeed lighter and also saves time but my concern is it ranges from very user unfriendly to mildly user unfriendly depending on the implementation of the method as opposed to a table which is easier on the eyes.

achobanov replied:
Do we still need this file? Why don't you replace it with `NotChildren`?

StoyanGoranov commented on src-v2/NTS.Domain.Setup/Entities/Loop.cs:
```diff
@@ -1,19 +1,63 @@
 
     public double Distance { get; private set; }
-	public int Rec { get; private set; }
+	public int Recovery { get; private set; }
```
Added a semantically complete property name because it makes the code more readable - when one reads it, it gives information about its purpose and avoids a certain level of ambiguity.

StoyanGoranov commented on src-v2/Judge/NTS.Judge.Blazor/Pages/Setup/Competitions/CompetitionForm.razor:
```diff
@@ -35,7 +35,7 @@
     {
         DateTime start = startDay.GetValueOrDefault(DateTime.MinValue);
-        DateTime startDayTime = start.Add((TimeSpan)startTime);
+        DateTime startDayTime = start.Date.Add((TimeSpan)startTime);
```
GetValueOrDefault is a nice way to avoid an error when the nullable values are null but it will break the functionality of the application since the values are used within the logic. I think it is best to throw an error if the respective inputs contain null values.

StoyanGoranov replied:
Unsure, which project do you reference by the name 'Not.Common'. Am I missing something?

StoyanGoranov commented on src-v2/Judge/NTS.Judge.Blazor/Pages/Setup/Competitions/CompetitionFormModel.cs:
```diff
@@ -16,13 +17,15 @@ public CompetitionFormModel()
+        Id = competition.Id;
         Name = competition.Name;
         Type = competition.Type;
         DateTime? startDay = competition.StartTime.DateTime;
         TimeSpan? startTime = competition.StartTime.DateTime.TimeOfDay;
         StartDay = startDay;
         StartTime = startTime;
```
Yes, this is part of the logic that separates the Date and the Time of the Competition in separate inputs. If removed the functionality breaks down.

StoyanGoranov committed:
address review

StoyanGoranov committed:
Add NotChildren Component

NotChildren replaces the individual ComponentGrids by injecting the same
interfaces and creating rows for the child items by utilizing the
ToString method of each DomainEntity.

https://github.com/Not-Endurance/not-timing-system/pull/195

StoyanGoranov committed:
Replace LoopGrid with new NotChildren Component

https://github.com/Not-Endurance/not-timing-system/pull/195

StoyanGoranov committed:
Improve NotChildren Component and CompetitionGrid appearance

Adding a MudPaper container to the NotChildren as well as some html
classes and styling at NTS.Judge.MAUI's app.css

achobanov commented on src-v2/Judge/NTS.Judge.Blazor/Pages/Setup/Competitions/CompetitionForm.razor:
```diff
@@ -26,24 +27,17 @@
-        DateTime startDayTime = start.Date.Add((TimeSpan)startTime);
-        DateTimeOffset startTimeOffset = ToDateTimeOffSet(startDayTime);
+        DateTime startDayTime = start.Date.Add(startTime.GetValueOrDefault(TimeSpan.Zero));
+        DateTimeOffset startTimeOffset = startDayTime.ToDateTimeOffset();
```
I think we should either keep them nullable or set default values on the model object.

StoyanGoranov replied:
Not sure I understand what you mean in this comment.

achobanov commented on src-v2/Judge/NTS.Judge.Blazor/Pages/Setup/Competitions/CompetitionUpdate.razor:
```diff
@@ -14,7 +15,10 @@
 </MudPaper>
 
-<LoopGrid />
+<NotChildren Name="@Localizer.Get("Loops")" T="Loop" TForm="LoopForm"
```
Localizer can be used internally in the component itself.

achobanov commented on src-v2/Judge/NTS.Judge.Blazor/Pages/Setup/Contestants/ContestantForm.razor:
```diff
@@ -3,29 +3,32 @@
 
 <EditForm Model="_model">
-    <MudTimePicker DisableToolbar="true" @ref="_timeField" @bind-Time="_model.StartTime" Label="@Localizer.Get("Time")" />
+    <MudSwitch T="bool" @ref="_unrankedField" @bind-Value="_model.Unranked" Label="@Localizer.Get("Unranked")"
+               ThumbIcon="@(_model.Unranked==true ? Icons.Material.Filled.Done : Icons.Material.Filled.Close)"
+               ThumbIconColor="@(_model.Unranked==true ? Color.Success : Color.Error)" />
+    <MudSwitch T="bool" @bind-Value="_overrideStartTime" Label="@Localizer.Get("Override Start Time")"
+               ThumbIcon="@(_overrideStartTime==true ? Icons.Material.Filled.Done : Icons.Material.Filled.Close)"
+               ThumbIconColor="@(_overrideStartTime==true ? Color.Success : Color.Error)" />
```
The switch configurations seem very similar. Can't we standardize them?

achobanov commented on src-v2/Judge/NTS.Judge.Blazor/Pages/Setup/Contestants/ContestantForm.razor:
```diff
@@ -35,13 +38,13 @@
     {
-        DateTimeOffset newStartTimeOffset = ToDateTimeOffSet(_model.StartTime);
-        return Contestant.Create(newStartTimeOffset);
+        DateTimeOffset? newStartTimeOffset = _model.StartTimeOverride.ToDateTimeOffset();
```
Use `var newStartTimeOffset` instead of declaring the type explicitly,

achobanov commented on src-v2/Judge/NTS.Judge.Blazor/Pages/Setup/Contestants/ContestantForm.razor:
```diff
@@ -35,13 +38,13 @@
     {
-        DateTimeOffset newStartTimeOffset = ToDateTimeOffSet(_model.StartTime);
-        return Contestant.Update(_model.Id ,newStartTimeOffset);
+        DateTimeOffset? newStartTimeOffset = _model.StartTimeOverride.ToDateTimeOffset();
```
Use implicit var declaration.

achobanov commented on src-v2/Judge/NTS.Judge.Blazor/Pages/Setup/Contestants/ContestantForm.razor:
```diff
@@ -35,13 +38,13 @@
-        DateTimeOffset newStartTimeOffset = ToDateTimeOffSet(_model.StartTime);
-        return Contestant.Update(_model.Id ,newStartTimeOffset);
+        DateTimeOffset? newStartTimeOffset = _model.StartTimeOverride.ToDateTimeOffset();
+        return Contestant.Update(_model.Id ,newStartTimeOffset,_model.Unranked);
```
Parameters should be separated by a comma and a space - `, `

achobanov commented on src-v2/Judge/NTS.Judge.Blazor/Pages/Setup/Contestants/ContestantFormModel.cs:
```diff
@@ -11,15 +11,24 @@ public class ContestantFormModel
-        StartTime = startTime;
+        if (contestant.StartTimeOverride != null)
+        {
+            TimeSpan? startTime = contestant.StartTimeOverride.Value.DateTime.TimeOfDay;
```
This isn't correct. I'm not sure exactly what time this will return, but probably it will  simply be the time of day for the time-zone it was created in. So if it was in 6:00 in Sofia time then it could simply return 6:00 when you open it up in UK. You need to convert the DateTimeOffset value to the current timezone. This isn't a big issue right now, but eventually - when we have web APIs and databases it could become one.

achobanov replied:
Also implicit vars. Explicit declaration is only acceptable if it is necessary.

achobanov commented on src-v2/Judge/NTS.Judge.Blazor/Pages/Setup/Contestants/ContestantFormModel.cs:
```diff
@@ -11,15 +11,24 @@ public class ContestantFormModel
+        if (contestant.StartTimeOverride != null)
+        {
+            TimeSpan? startTime = contestant.StartTimeOverride.Value.DateTime.TimeOfDay;
+            StartTimeOverride = startTime;
```
Also no need to create a variable and then assign it to the property. simply do
```
StartTimeOverride = contestant.StartTimeOverride....;
```

achobanov commented on src-v2/Judge/NTS.Judge.Blazor/Pages/Setup/Contestants/ContestantFormModel.cs:
```diff
@@ -11,15 +11,24 @@ public class ContestantFormModel
+            TimeSpan? startTime = contestant.StartTimeOverride.Value.DateTime.TimeOfDay;
+            StartTimeOverride = startTime;
+        }
+        else
+        {
+            StartTimeOverride = null;
+        }
```
Nullable definitions contain null value by default, it's how the runtime works, this expression is completely redundant.

achobanov commented on src-v2/Judge/NTS.Judge.Blazor/Pages/Setup/Contestants/ContestantFormModel.cs:
```diff
@@ -11,15 +11,24 @@ public class ContestantFormModel
     public int Id { get; set; }
-    public TimeSpan? StartTime { get; set; }
+    public TimeSpan? StartTimeOverride { get; set; }
+    public Boolean  Unranked { get; set; }
```
Bool properties should indicate it in their name by posing a question. `IsUnranked` or `IsNotRanked` would be better.

achobanov commented on src-v2/Judge/NTS.Judge.Blazor/Pages/Setup/Events/EventPage.razor:
```diff
@@ -35,7 +35,7 @@
         var entity = await _behind.Read(default);
         if (entity == null)
         {
-            throw GuardHelper.Exception($"Event cannot be null here");
+           GuardHelper.ThrowIfNull(entity, "Event cannot be null here");    
         }
```
Does this make sense to you? Did you look inside the definition of `ThrowIfNull`?

achobanov commented on src-v2/Judge/NTS.Judge.MAUI/wwwroot/css/app.css:
```diff
@@ -3,6 +3,19 @@
     margin-bottom: 5px;
 }
 
+.action-column{
+    display:inline;
+    margin-left:20px;
+}
+
+.half-width{
+    width:45%;
+}
+
+.children-component{
+    margin-top:10px;
+}
+
```
Can't we use Mud's styling framework instead of custom styles?

StoyanGoranov replied:
Some styles can be set with Mud styling.
Like `margin-top:10px;` but the general idea here is to set a class which can target an entire component or apply styles that are missing from Mud as a semantic property, e.g. `half-width` should ensure that two components with the class will fit on the same line.
I suggest we keep these class configurations until the focus comes to the UI fine tuning.

achobanov commented on src-v2/Judge/NTS.Judge/Events/CompetitionChildrenBehind.cs:
```diff
@@ -9,18 +9,17 @@ public class CompetitionChildrenBehind : INotBehindParent<Contestant>, INotBehin
     private readonly IParentRepository<Contestant> _contestantParentRepository;
     private readonly IParentRepository<Loop> _loopParentRepository;
-    private Competition? _competition;
+    public Competition? _competition;
```
Why public? Fields should always be private. Underscore also should only prefix private fields/properties.

achobanov commented on src-v2/Judge/NTS.Judge/Events/ContestantChildrenBehind.cs:
```diff
@@ -3,7 +3,7 @@
 namespace NTS.Judge.Events;
 
-public class ContestantChildrenBehind : INotBehindWithChildren<Contestant>
+public class ContestantChildrenBehind 
```
Do we need this class?

achobanov commented on src-v2/NTS.Domain.Setup/Entities/Competition.cs:
```diff
@@ -14,34 +14,20 @@ public class Competition : DomainEntity, ISummarizable, IParent<Contestant>, IPa
+        if (type == default)
         {
-            throw new DomainException(nameof(Type), "Competition type must have a value different from 0.");
+            throw new DomainException(nameof(Type), "Competition Type cannot be default");
```
This is far better than "different from 0", but why not "Competition type is required"?

achobanov commented on src-v2/NTS.Domain.Setup/Entities/Contestant.cs:
```diff
@@ -6,37 +6,32 @@
     {
-
-        if (startTimeOverride.DateTime.CompareTo(DateTime.Today) < 0)
+        if (startTimeOverride!=null && startTimeOverride.Value.DateTime.CompareTo(DateTime.Today) < 0)
```
Space around operators.

achobanov commented on src-v2/NTS.Domain.Setup/Entities/Loop.cs:
```diff
@@ -5,55 +5,33 @@ namespace NTS.Domain.Setup.Entities;
 public class Loop : DomainEntity, ISummarizable
 {
-    public static Loop Create(double distance, int rec, int rest, bool isFinal) => new(distance, rec, rest, isFinal);
+    public static Loop Create(double distance, int rec, int rest) => new(distance, rec, rest);
```
rec -> recovery

achobanov commented on src-v2/NTS.Domain.Setup/Entities/Loop.cs:
```diff
@@ -5,55 +5,33 @@ namespace NTS.Domain.Setup.Entities;
+    public static Loop Create(double distance, int rec, int rest) => new(distance, rec, rest);
 
-    public static Loop Update(int id, double distance, int rec, int rest, bool isFinal) => new(id, distance, rec, rest, isFinal);
+    public static Loop Update(int id, double distance, int rec, int rest) => new(id, distance, rec, rest);
```
rec -> recovery

achobanov commented on src-v2/NTS.Domain.Setup/Entities/Official.cs:
```diff
@@ -17,9 +17,9 @@ private Official(int id, Person person, OfficialRole role)
+        if (role == default)
         {
-            throw new DomainException(nameof(Role), "Official Role must have a value different from 0.");
+            throw new DomainException(nameof(Role), "Official Role cannot be default");
```
required

achobanov commented on src-v2/Not/Not.Blazor/TM/NotChildren.razor:
```diff
@@ -0,0 +1,62 @@
+    @foreach (var item in Items)
+    {
+        <p class="mt-1 ml-6 mb-1">
+            @item.ToString() 
```
Think `ToString()` will be called implicitly by the blazor framework. Try it and remove it if isn't necessary.

achobanov commented on src-v2/Not/Not.Blazor/TM/NotChildren.razor:
```diff
@@ -0,0 +1,62 @@
+    [Parameter]
+    public string Name { get; set; } = default!;
+    [Parameter]
+    public string NoItemsMessage { get; set; } = default!;
```
This message is going to be the same for every entity. Why not simply construct it using the `Name` parameter and a string constant?

StoyanGoranov replied:
Kept the parameter for custom messages but now setting the NoItemsMessage as a default.

achobanov commented on src-v2/Not/Not.Blazor/TM/NotChildren.razor:
```diff
@@ -0,0 +1,62 @@
+    private INotBehindParent<T> _parentBehind { get; set; } = default!;
+    [Inject]
+    private IFormNavigator<T, TForm> _contestantNavigator { get; set; } = default!;
+    [Inject]
+    private INavigator _navigator { get; set; } = default!;
```
Is this used?

achobanov commented on src-v2/Not/Not/Extensions/DateTimeOffsetExtension.cs:
```diff
@@ -0,0 +1,25 @@
+        DateTimeOffset offsetTime = timeWithSpecifiedKind;
+        return offsetTime;
+    }
+    public static DateTimeOffset? ToDateTimeOffset(this TimeSpan? timeToBeAdded)
```
I don't think we should involve nullability in extensions, unless we're extending `Nullable` itself. Instead you should make sure the method is not executed on null value. See [null-conditional operator](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/member-access-operators#null-conditional-operators--and-)

achobanov reviewed (commented):
Overall pretty good progress in the right direction. Mostly tiding up to do :) Good job

achobanov reviewed (changes requested):
Still required tho :P

StoyanGoranov committed:
Replace ContestantGrid at CompetitionUpdate with NotChildren

Utilize the new NotChildren component for the ContestantGrid as well and
improve the overall alignment of the Html elements.

https://github.com/Not-Endurance/not-timing-system/pull/195

StoyanGoranov committed:
Add required changes after peer review

https://github.com/Not-Endurance/not-timing-system/pull/195

StoyanGoranov committed:
Implement NotSwitch component and replace it at ContestantForm

The NotSwitch uses a MudBlazor Switch with default values.

https://github.com/Not-Endurance/not-timing-system/pull/195

achobanov commented on src-v2/Judge/NTS.Judge.Blazor/Pages/Setup/Competitions/CompetitionUpdate.razor:
```diff
@@ -10,17 +10,20 @@
-
-<ContestantGrid />
+<MudPaper Class="d-flex justify-space-between pr-6" Elevation="0">
+    <NotChildren Name="Loops" T="Loop" TForm="LoopForm"
+        UpdateRoute="@LOOP_UPDATE">
+    </NotChildren>
```
When defined like this the component looks as if UpdateRoute is it's content. Why is it on a new line and why not self-closing component?

achobanov commented on src-v2/Judge/NTS.Judge.Blazor/Pages/Setup/Contestants/ContestantFormModel.cs:
```diff
@@ -18,17 +18,12 @@ public ContestantFormModel(Contestant contestant)
-            TimeSpan? startTime = contestant.StartTimeOverride.Value.DateTime.TimeOfDay;
-            StartTimeOverride = startTime;
+            StartTimeOverride = contestant.StartTimeOverride.Value.LocalDateTime.TimeOfDay;
         }
```
Much better, but I think it can still be simplified:
```
StartTimeOverride = contestant.StartTimeOverride?.LocalDateTime.TimeOfDay
```

achobanov commented on src-v2/Judge/NTS.Judge.Blazor/Pages/Setup/Events/EventPage.razor:
```diff
@@ -35,7 +35,7 @@
         {
-           GuardHelper.ThrowIfNull(entity, "Event cannot be null here");    
+           GuardHelper.ThrowIfNull(entity);    
         }
```
Haha, not quite. Take a careful look at lines I've marked here and then look at the definition of `ThrowIfNull` again. If you still can't see it - replace the invocation of `ThrowIfNull`  with it's definition and you'll see what I mean.

achobanov commented on src-v2/NTS.Domain.Setup/Entities/Contestant.cs:
```diff
@@ -20,18 +20,18 @@ private Contestant(int id, DateTimeOffset? startTimeOverride, bool unranked) : t
         }
         StartTimeOverride = startTimeOverride;
-        Unranked = unranked;
+        IsUnranked = unranked;
```
Rename variable name as well

achobanov commented on src-v2/Not/Not.Blazor/TM/NotChildren.razor:
```diff
@@ -45,18 +47,25 @@ else
     [Parameter]
     public string UpdateRoute { get; set; } = default!;
+    [Parameter]
+    public string Class { get; set; } = default!;
```
Do you think there is value in exposing parameters that aren't being used? It's not a hard thing to add a parameter if you need it at some point. And allowing unnecessary configuration can lead to overengineering and inconsistencies, both of which aren't ideal. Same goes for the `NoItemsMessage`.

achobanov commented on src-v2/Not/Not.Blazor/TM/NotChildren.razor:
```diff
@@ -45,18 +47,25 @@ else
     protected override void OnInitialized()
     {
+        Name = Localizer.Get(Name);
+        if(NoItemsMessage == default)
```
I would rename to the simpler `EmptyMessage`.

achobanov commented on src-v2/Not/Not.Blazor/TM/NotChildren.razor:
```diff
@@ -45,18 +47,25 @@ else
+        Name = Localizer.Get(Name);
+        if(NoItemsMessage == default)
+        {
+            string defaultMessage = $"No {Name.ToLower()} have been created for this event.";
+            NoItemsMessage = Localizer.Get(defaultMessage);
```
When defined like this you will have to add multiple entries in the localization resex files - one for every usage of `NotChildren` with a different `Name`. Instead you could define the property as a constant that contains a string placeholder and use `string.Format`. That way we will have to translate a single resex entry.  However this isn't currently supported by the `Ilocalizer` so you can just leave a TODO comment so we don't forget about it later.

achobanov commented on src-v2/Not/Not/Extensions/DateTimeOffsetExtension.cs:
```diff
@@ -13,11 +14,10 @@ public static DateTimeOffset ToDateTimeOffset(this DateTime time)
-        var time = today.Add((TimeSpan)timeToBeAdded);
+        DateTime time = today.Add(timeToBeAdded);
         DateTime timeWithSpecifiedKind = DateTime.SpecifyKind(time, DateTimeKind.Local);
         DateTimeOffset offsetTime = timeWithSpecifiedKind;
```
Implicit vars again, It seems you have a habit of using them (or is it the IDE somehow? ). Try to be more vigilant and replace then with `var` whenever you see them, until you make a habit of using it in the first place.

StoyanGoranov committed:
Improve logic and cleanup code

Address review

https://github.com/Not-Endurance/not-timing-system/pull/195

StoyanGoranov committed:
Rename NotChildren NoItemsMessage to EmptyMessage

https://github.com/Not-Endurance/not-timing-system/pull/195

achobanov reviewed (changes requested):
Looks good, but I've found a bug with Contestant Update - StartTimeOverride is kept, but  it's checkbox is always off. which isnt correct. Also do you think we should add `StartTimeOverride` and `IsUnranked` in ToString?

StoyanGoranov committed:
Fix ContestantForm Update StartTimeOverride Switch Bug

The fix involves an additional OnChange Event implementation in the
NotSwitch which fires to apply .hidden and .visible Html classes.
The Html classes are set by a SetVisibilityClass method.
The styling is added at the app.css.

https://github.com/Not-Endurance/not-timing-system/pull/195

StoyanGoranov committed:
fix them all
